### ISSUE TITLE
Compiler error in vasprintf.c 

### ISSUE THREAD

**Issue by bradsen** [BrianGladman/mpfr#22]
![Screenshot (1324)](https://user-images.githubusercontent.com/38464644/90307256-003d3d80-df07-11ea-85e5-f82ddbfb3d7e.png)

Hi Brian, thanks for the codes. I have issue compiling vasprintf.c. There are 270 errors in total.  I do not know where to start debugging. Does the codes have any issues right from the start?

**Comment by BrianGladman** [comment]
There are no known errors in this project and it builds fine for me so we need to find out what is different in your system.  The error window you give is not especially helpful as it does not give the line numbers where the errors occur.  The full text of the output window would be more useful.   Do all other source files compile without any errors?

**Comment by bradsen** [comment]
![Screenshot (1334)](https://user-images.githubusercontent.com/38464644/90309495-ad21b580-df1b-11ea-849c-cfe604e07700.png)
Are you referring to this? There were 1 warning and 1 error.

For 'va_copy' redifinition error, I added the following:

#ifdef va_copy
#undef va_copy
#endif

Then I realised there were more errors. I'm able to run the mpir file. No errors at all. Then I link the mpfr to mpir file directory. Like what you mentioned previously. It works. Just that there are 2 errors.

**Comment by BrianGladman** [comment]
No, I am referring to the text in the output window when the 270 errors occur - this gives the details of the errors in vasprintf.c and their line numbers.   Is vasprintf.c the only file that fails to compile?

**Comment by bradsen** [comment]
Yup, vasprintf.c is the only file that fails to compile. 

Let me send you a few screenshots. Not sure if that is what you are looking for. vasprintf is exactly the same as the one in the repository. The only 2 settings that I did for visual studio, one is adding additional file directory, the other one is change the windows SDK version. I followed the tutorial guide created by Thomas Kim. 

https://www.youtube.com/watch?v=SoBVqS2lN7E


[Errors.pdf](https://github.com/BrianGladman/mpfr/files/5078668/Errors.pdf)

**Comment by BrianGladman** [comment]
Don't worry about the other errors until you find out why _MPFR_PREC_FORMAT is not being defined.  This is almost certain to be the cause of the problem.    It's hard to see how this symbol could be undefined so it might be worth temporarily adding mpfr.h as an include file in vasprintf.c to see if the error goes away.  This would at least identify the cause of the problem.

**Comment by 7ofNine** [comment]
Are you sure that you have the correct version of vasprintf.c ? Looking at your Errors.pdf I see the #error macro on line #108 while the latest version has it on line #103 and looking back in the repository I don't think that it ever has been on line #108.
Have a look if git has it in the index as changed.

**Comment by bradsen** [comment]
Thanks Brian and 7ofNine! 

I managed to solved one of the errors. I have copy and paste the correct version of vasprint.c and changed the directory in properties --> C/C++ --> General ---> Additional Include Directories and selected mpir file location. 

I still got 1 more error that I'm still trying to figure out. I have created a file in that specific location. But the program still return this error. Not sure why. 

![Screenshot (1348)](https://user-images.githubusercontent.com/38464644/90325906-24a32380-dfb4-11ea-8e22-b6ad6af31cc4.png)

**Comment by BrianGladman** [comment]
@7ofNine Thank you for spotting the issue with vasprintf.c
@bradsen  If you download the code in this repository and the MPIR repository correctly and put these two directories in a common root directory, both applications should build without the need to make any changes to the code or to the Visual Studio directory properties.  Since you need to cut and paste code for vasprintf.c and also change directory properties, it seems clear that you are not building MPFR in the way that this repository is designed to support.   I can help you to build if you use the same approach as I do, but not if you are using a different approach (as appears to be the case).

**Comment by bradsen** [comment]
I see. Thanks for the input Brian. How should I build MPFR by default?

**Comment by BrianGladman** [comment]
First, clone the mpir and mpfr repositories into the same location (i.e. the same directory).    

Second, download the mpir documentation here: http://mpir.org/mpir-3.0.0.pdf

Third, build your chosen version of MPIR by carefully following the instructions in the MPIR documentation. 

Fourth, build MPFR.   Although the readme.txt file in the repository is now out-of-date, it should be enough to indicate how to do this.

Ideally, don't change any Visual Studio properties unless you have to.  You should be OK changing the SDK and the build toolset, but don't change any other properties until you have built them first with only these changes.

**Comment by bradsen** [comment]
I have tried to follow the instructions in the MPIR documentation. But it doesn't seems like the steps are applicable. First off, I can't find ./configure inside the mpir file. I can't run it at all. Or even make, make check or make install.

**Comment by BrianGladman** [comment]
If you wish to build MPIR with Microsoft Visual Studio use section 2.4 in the documentation entitled "Building with Microsoft Visual Studio'.  It is listed in the Index and starts half way down page 11.
